Add outlier filter function

mastercaution opened the pull request:
A function to filter out potentially vulnerable samples.

Todo:
- [x] Improve doc
- [x] Change search algorithm that also works with other metrics than `cosine`

mastercaution committed:
Add outlier filter function

A function to filter out potentially vulnerable samples.

mastercaution committed:
Add GMIA references to filter documentation

hallojs commented on pepr/utilities.py:
```diff
@@ -51,3 +55,270 @@ def assign_record_ids_to_target_models(
+    # Step 8
+    filtered_target_indices = np.delete(data_conf["evaluation_indices"], target_records)
+
+    return filtered_target_indices, neighbor_threshold, probability_threshold
```
Probably we should rename ``filtered_target_indices`` to ``filtered_data_indices``?

mastercaution replied:
Renamed in c3a67ed

hallojs commented on pepr/utilities.py:
```diff
@@ -51,3 +55,270 @@ def assign_record_ids_to_target_models(
+        Dataset with all training samples used in the given pentesting setting.
+    labels : numpy.ndarray
+        Array of all labels used in the given pentesting setting.\
+    filter_pars : dict
+        Dictionary containing needed filter parameters:
+
+        * number_reference_models (int): Number of reference models to be trained.
+        * reference_training_set_size (int): Size of the trainings set for each
+          reference model.
+        * create_compile_model (function): Function that returns a compiled
+          TensorFlow model (typically identical to the target model) used in the
+          training of the reference models.
+        * reference_epochs (int): Number of training epochs of the reference models.
+        * reference_batch_size (int): Batch size used in the training of the
+          reference models.
+        * hlf_metric (str): Metric (typically 'cosine') used for the distance
+          calculations in the high-level feature space. For valid metrics see
+          documentation of sklearn.metrics.pairwise_distances.
+        * hlf_layer_number (int): If value is n, the n-th layer of the model
+          returned by create_compile_model is used to extract the high-level feature
+          vectors.
+        * neighbor_threshold (float): If distance is smaller then the neighbor
+          threshold the record is selected as target record.
+        * probability_threshold (float): For details see section 4.3 from the paper.
+        * number_target_records (int): If set, the selection algorithm performs
+          `max_search_rounds`, to find a `neighbor_threshold`, that leads to a finding
+          of `n_targets` target records. These target records are most vulnerable with
+          respect to our selection criterion.
+        * max_search_rounds (int): If `number_target_records` is given, maximal
+          `max_search_rounds` are performed to find `number_target_records` of potential
+          vulnerable target records.
```
I think ``number_classes``is missed here? :)

mastercaution replied:
Fixed in c3a67ed

mastercaution committed:
Fix filter doc and variable name

mastercaution committed:
Remove `data_conf` parameter from filter function

`data_conf` is not needed for filtering only, so it was removed.
Therefore a custom search for the outliers was added that works on the
entire dataset given.

mastercaution committed:
Improve binary search to work with every metric

mastercaution committed:
Add search area explanation and fix code style

hallojs commented on pepr/utilities.py:
```diff
@@ -51,3 +55,282 @@ def assign_record_ids_to_target_models(
+    3. Generate intermediate models.
+    4. Extract high-level features.
+    5. Compute pairwise distances between high-level features.
+    6. Determine potential vulnerable records.
+    7. Remove potential vulnerable records.
```
The function name is ``filter_out_outlier``so lets change the step descriptions accordingly:
6. Determine outlier records.
7. Remove outlier records from the dataset.

Probably we also want to rename the function to ``filter_outlier``, that's a little cleaner? :)

mastercaution replied:
Done in 30c739a

hallojs commented on pepr/utilities.py:
```diff
@@ -51,3 +55,282 @@ def assign_record_ids_to_target_models(
+    filter_pars : dict
+        Dictionary containing needed filter parameters:
+
+        * number_classes (int): Number of different classes the target model
+          predict.
```
... number of different classes in the dataset?

mastercaution replied:
Changed in 30c739a

hallojs commented on pepr/utilities.py:
```diff
@@ -51,3 +55,282 @@ def assign_record_ids_to_target_models(
+        * number_reference_models (int): Number of reference models to be trained.
+        * reference_training_set_size (int): Size of the trainings set for each
+          reference model.
+        * create_compile_model (function): Function that returns a compiled
+          TensorFlow model (typically identical to the target model) used in the
+          training of the reference models.
```
...(in gmia this is typically...)...

hallojs commented on pepr/utilities.py:
```diff
@@ -51,3 +55,282 @@ def assign_record_ids_to_target_models(
+          threshold the record is selected as target record.
+        * number_neighbor_threshold (float): If number of neighbors of a record is
+          smaller than this, it is considered a vulnerable example.
+        * number_outlier (int): If set, the selection algorithm performs
+          `max_search_rounds`, to find a `distance_neighbor_threshold`, that leads to a
+          finding of `n_targets` target records. These target records are most
+          vulnerable with respect to our selection criterion.
```
I think if you like you can move your nice description of the initialization of the ``neighbor_threshold``to this bullet point 👍

mastercaution replied:
I think it's an implementation detail that does not need to be in the documentation

hallojs reviewed (changes requested):
I think after adapting the few doctoring lines we can merge this one. Looks great!

mastercaution committed:
Improve `filter_outlier` docstrings
